File: naro-backend/app/schemas/tow.py
"""Tow Pydantic schemas — Zod parity (packages/domain/src/tow.ts).

Backend bazı alanlarda Zod'dan daha geniş (preauth_failed, preauth_stale stages;
preauth_stale + kasko_rejected settlement statuses; required_equipment array).
Schema parity CI testi bu farkları known-drift olarak işaretler.
"""

from __future__ import annotations

from datetime import datetime
from decimal import Decimal
from enum import StrEnum
from typing import Annotated, Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

# ─── Enums (StrEnum — JSON serde ile uyumlu) ────────────────────────────────


class TowModeSchema(StrEnum):
    IMMEDIATE = "immediate"
    SCHEDULED = "scheduled"


class TowEquipmentSchema(StrEnum):
    FLATBED = "flatbed"
    HOOK = "hook"
    WHEEL_LIFT = "wheel_lift"
    HEAVY_DUTY = "heavy_duty"
    MOTORCYCLE = "motorcycle"


class TowIncidentReasonSchema(StrEnum):
    NOT_RUNNING = "not_running"
    ACCIDENT = "accident"
    FLAT_TIRE = "flat_tire"
    BATTERY = "battery"
    FUEL = "fuel"
    LOCKED_KEYS = "locked_keys"
    STUCK = "stuck"
    OTHER = "other"


class TowDispatchStageSchema(StrEnum):
    SEARCHING = "searching"
    ACCEPTED = "accepted"
    EN_ROUTE = "en_route"
    NEARBY = "nearby"
    ARRIVED = "arrived"
    LOADING = "loading"
    IN_TRANSIT = "in_transit"
    DELIVERED = "delivered"
    CANCELLED = "cancelled"
    TIMEOUT_CONVERTED_TO_POOL = "timeout_converted_to_pool"
    SCHEDULED_WAITING = "scheduled_waiting"
    BIDDING_OPEN = "bidding_open"
    OFFER_ACCEPTED = "offer_accepted"
    PREAUTH_FAILED = "preauth_failed"
    PREAUTH_STALE = "preauth_stale"


class TowSettlementStatusSchema(StrEnum):
    NONE = "none"
    PRE_AUTH_HOLDING = "pre_auth_holding"
    PREAUTH_STALE = "preauth_stale"
    FINAL_CHARGED = "final_charged"
    REFUNDED = "refunded"
    CANCELLED = "cancelled"


# ─── B-P1-4: tow stage label + phase projection ─────────────────────────


# FE UX: tow shell coarse status ("matching" vs "service_in_progress"),
# stage detay burada. FE tow_phase'e göre ana chip, stage_label alt-satır.
_STAGE_META: dict[TowDispatchStageSchema, tuple[str, str]] = {
    TowDispatchStageSchema.SEARCHING: (
        "Çekici aranıyor",
        "dispatch",
    ),
    TowDispatchStageSchema.TIMEOUT_CONVERTED_TO_POOL: (
        "Havuza gönderildi",
        "dispatch",
    ),
    TowDispatchStageSchema.SCHEDULED_WAITING: (
        "Planlı çekici bekliyor",
        "bidding",
    ),
    TowDispatchStageSchema.BIDDING_OPEN: (
        "Teklifler alınıyor",
        "bidding",
    ),
    TowDispatchStageSchema.OFFER_ACCEPTED: (
        "Teklif kabul edildi",
        "bidding",
    ),
    TowDispatchStageSchema.ACCEPTED: (
        "Çekici atandı",
        "en_route",
    ),
    TowDispatchStageSchema.EN_ROUTE: (
        "Çekici yolda",
        "en_route",
    ),
    TowDispatchStageSchema.NEARBY: (
        "Çekici yakında",
        "en_route",
    ),
    TowDispatchStageSchema.ARRIVED: (
        "Çekici vardı",
        "on_scene",
    ),
    TowDispatchStageSchema.LOADING: (
        "Araç yükleniyor",
        "on_scene",
    ),
    TowDispatchStageSchema.IN_TRANSIT: (
        "Araç taşınıyor",
        "delivery",
    ),
    TowDispatchStageSchema.DELIVERED: (
        "Araç teslim edildi",
        "delivery",
    ),
    TowDispatchStageSchema.CANCELLED: (
        "Çekici iptal edildi",
        "delivery",
    ),
    TowDispatchStageSchema.PREAUTH_FAILED: (
        "Ödeme başarısız",
        "dispatch",
    ),
    TowDispatchStageSchema.PREAUTH_STALE: (
        "Ödeme tekrar onayı gerek",
        "dispatch",
    ),
}


def tow_stage_label(stage: TowDispatchStageSchema) -> str:
    return _STAGE_META.get(stage, ("Bilinmeyen aşama", "dispatch"))[0]


def tow_phase(stage: TowDispatchStageSchema) -> str:
    return _STAGE_META.get(stage, ("", "dispatch"))[1]


class TowEvidenceKindSchema(StrEnum):
    CUSTOMER_PRE_STATE = "customer_pre_state"
    TECH_ARRIVAL = "tech_arrival"
    TECH_LOADING = "tech_loading"
    TECH_DELIVERY = "tech_delivery"


class TowDispatchResponseSchema(StrEnum):
    PENDING = "pending"
    ACCEPTED = "accepted"
    DECLINED = "declined"
    TIMEOUT = "timeout"


# ─── Primitives ─────────────────────────────────────────────────────────────


class LatLng(BaseModel):
    model_config = ConfigDict(extra="forbid")

    lat: Annotated[float, Field(ge=-90, le=90)]
    lng: Annotated[float, Field(ge=-180, le=180)]


# ─── Value objects ──────────────────────────────────────────────────────────


class TowFareQuote(BaseModel):
    """Hemen/randevulu fare hesap snapshot — cap/locked price dahil."""

    model_config = ConfigDict(extra="forbid")

    mode: TowModeSchema
    base_amount: Decimal
    distance_km: Decimal
    per_km_rate: Decimal
    urgency_surcharge: Decimal
    buffer_pct: Decimal
    cap_amount: Decimal
    locked_price: Decimal | None = None
    currency: str = "TRY"


class TowLiveLocationSchema(BaseModel):
    case_id: UUID
    technician_id: UUID
    lat: Annotated[float, Field(ge=-90, le=90)]
    lng: Annotated[float, Field(ge=-180, le=180)]
    heading: Annotated[float, Field(ge=0, le=360)] | None = None
    speed_kmh: Annotated[float, Field(ge=0)] | None = None
    captured_at: datetime


class TowKaskoDeclaration(BaseModel):
    model_config = ConfigDict(extra="forbid")

    has_kasko: bool = False
    insurer_name: str | None = None
    policy_number: str | None = None
    pre_auth_on_customer_card: bool = True


class TowDispatchAttemptSchema(BaseModel):
    id: UUID
    technician_id: UUID
    technician_name: str | None = None
    attempt_order: int
    sent_at: datetime
    response_at: datetime | None = None
    response: TowDispatchResponseSchema = TowDispatchResponseSchema.PENDING
    distance_km: Decimal | None = None
    eta_minutes: int | None = None


class TowTechnicianProfileSchema(BaseModel):
    id: UUID
    name: str
    rating: float
    completed_jobs: int
    plate: str | None = None
    truck_model: str | None = None
    equipment: TowEquipmentSchema
    phone: str | None = None
    photo_url: str | None = None


class TowBid(BaseModel):
    id: UUID
    technician: TowTechnicianProfileSchema
    price_amount: Decimal
    price_label: str
    eta_window_label: str
    equipment: TowEquipmentSchema
    guarantee_label: str | None = None
    submitted_at: datetime


class TowEvidenceSchema(BaseModel):
    id: UUID
    case_id: UUID
    kind: TowEvidenceKindSchema
    uploader: Literal["customer", "technician", "system"]
    photo_url: str
    caption: str | None = None
    created_at: datetime


class TowOtpChallenge(BaseModel):
    code: str
    purpose: Literal["arrival", "delivery"]
    recipient: Literal["customer", "delivery_recipient"]
    issued_at: datetime
    expires_at: datetime
    verified_at: datetime | None = None


# ─── Request / response ─────────────────────────────────────────────────────


class TowFareQuoteRequest(BaseModel):
    """POST /tow/fare/quote input."""

    model_config = ConfigDict(extra="forbid")

    mode: TowModeSchema
    pickup_lat_lng: LatLng
    dropoff_lat_lng: LatLng | None = None
    required_equipment: list[TowEquipmentSchema] = Field(default_factory=list)
    urgency_bump: bool = False


class TowCreateCaseRequest(BaseModel):
    """POST /tow/cases input (Zod TowRequest + fare quote)."""

    model_config = ConfigDict(extra="forbid")

    mode: TowModeSchema
    pickup_lat_lng: LatLng
    pickup_label: str
    dropoff_lat_lng: LatLng | None = None
    dropoff_label: str | None = None
    vehicle_id: UUID
    incident_reason: TowIncidentReasonSchema
    required_equipment: list[TowEquipmentSchema] = Field(default_factory=list)
    scheduled_at: datetime | None = None
    fare_quote: TowFareQuote
    kasko: TowKaskoDeclaration = Field(default_factory=TowKaskoDeclaration)
    attachments: list[UUID] = Field(default_factory=list)
    # Faz 2: accident/breakdown parent → tow çocuk bağlantısı (opsiyonel).
    # Müşteri mevcut kaza/arıza case'inden çekici çağırırsa doldurulur.
    parent_case_id: UUID | None = None


class TowDispatchResponseInput(BaseModel):
    """POST /tow/cases/{id}/dispatch/response — teknisyen accept/decline."""

    model_config = ConfigDict(extra="forbid")

    attempt_id: UUID
    response: Literal["accepted", "declined"]
    rejection_reason: str | None = None


class TowLocationInput(BaseModel):
    """POST /tow/cases/{id}/location — GPS ping."""

    model_config = ConfigDict(extra="forbid")

    lat: Annotated[float, Field(ge=-90, le=90)]
    lng: Annotated[float, Field(ge=-180, le=180)]
    heading_deg: Annotated[int, Field(ge=0, le=360)] | None = None
    speed_kmh: Annotated[int, Field(ge=0, le=300)] | None = None
    accuracy_m: Annotated[int, Field(ge=0, le=5000)] | None = None
    captured_at: datetime


class TowOtpIssueInput(BaseModel):
    model_config = ConfigDict(extra="forbid")
    purpose: Literal["arrival", "delivery"]
    recipient: Literal["customer", "delivery_person"]


class TowOtpVerifyInput(BaseModel):
    model_config = ConfigDict(extra="forbid")
    purpose: Literal["arrival", "delivery"]
    code: Annotated[str, Field(min_length=4, max_length=8)]


class TowCancelInput(BaseModel):
    model_config = ConfigDict(extra="forbid")
    reason_code: str
    reason_note: str | None = None


class TowKaskoDeclareInput(BaseModel):
    model_config = ConfigDict(extra="forbid")
    declaration: TowKaskoDeclaration


class TowRatingInput(BaseModel):
    model_config = ConfigDict(extra="forbid")
    rating: Annotated[int, Field(ge=1, le=5)]
    review_note: str | None = None


class TowBidSubmitInput(BaseModel):
    """POST /tow/bids — scheduled mode usta teklifi."""

    model_config = ConfigDict(extra="forbid")

    case_id: UUID
    price_amount: Decimal
    eta_minutes: int
    equipment: TowEquipmentSchema
    guarantee_label: str | None = None


# ─── Responses ──────────────────────────────────────────────────────────────


class TowFareQuoteResponse(BaseModel):
    quote: TowFareQuote
    pickup_address: str | None = None
    dropoff_address: str | None = None
    distance_km: Decimal
    expires_at: datetime


class TowCaseSnapshot(BaseModel):
    """GET /tow/cases/{id} response."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    created_at: datetime
    updated_at: datetime
    mode: TowModeSchema
    stage: TowDispatchStageSchema
    # B-P1-4 fix: shell coarse status scheduled tow'da "matching" görünür
    # (BIDDING_OPEN stage, shell sync coarse MATCHING). FE sub-label +
    # phase ile gerçek tow evresini render eder — shell status'una
    # takılmadan.
    stage_label: str
    tow_phase: str
    status: str
    # P1-G fix (QA tur 1): bidirectional link — accident/breakdown parent
    # case'i tow-tarafından navigate için döner (Faz 2 tow_case.parent_case_id
    # DB'de zaten var, serializer atlıyordu).
    parent_case_id: UUID | None = None
    pickup_lat_lng: LatLng | None
    pickup_label: str | None
    dropoff_lat_lng: LatLng | None
    dropoff_label: str | None
    incident_reason: TowIncidentReasonSchema | None
    required_equipment: list[TowEquipmentSchema]
    scheduled_at: datetime | None
    fare_quote: TowFareQuote | None
    assigned_technician_id: UUID | None
    settlement_status: TowSettlementStatusSchema
    final_amount: Decimal | None
    cancellation_fee: Decimal | None


class TowTrackingSnapshot(BaseModel):
    """GET /tow/cases/{id}/tracking — WS fallback polling."""

    case_id: UUID
    stage: TowDispatchStageSchema
    technician_id: UUID | None
    last_location: LatLng | None
    last_location_at: datetime | None
    eta_minutes: int | None
    updated_at: datetime


class TowDispatchResponseOutput(BaseModel):
    attempt_id: UUID
    response: TowDispatchResponseSchema
    next_stage: TowDispatchStageSchema | None

File: packages/domain/src/tow.ts
import { z } from "zod";

import { LatLngSchema } from "./technician";

// ───────── Enums ─────────

export const TowServiceModeSchema = z.enum(["immediate", "scheduled"]);
export type TowServiceMode = z.infer<typeof TowServiceModeSchema>;

export const TowVehicleEquipmentSchema = z.enum([
  "flatbed",
  "hook",
  "wheel_lift",
  "heavy_duty",
  "motorcycle",
]);
export type TowVehicleEquipment = z.infer<typeof TowVehicleEquipmentSchema>;

export const TowIncidentReasonSchema = z.enum([
  "not_running",
  "accident",
  "flat_tire",
  "battery",
  "fuel",
  "locked_keys",
  "stuck",
  "other",
]);
export type TowIncidentReason = z.infer<typeof TowIncidentReasonSchema>;

// Canonical BE (tow-priority audit 2026-04-23 P1-4 + matching-audit P2-2):
// preauth_failed + preauth_stale değerleri shared domain'e eklendi.
export const TowDispatchStageSchema = z.enum([
  "payment_required",
  "searching",
  "accepted",
  "en_route",
  "nearby",
  "arrived",
  "loading",
  "in_transit",
  "delivered",
  "cancelled",
  "timeout_converted_to_pool",
  "scheduled_waiting",
  "bidding_open",
  "offer_accepted",
  "preauth_failed",
  "preauth_stale",
]);
export type TowDispatchStage = z.infer<typeof TowDispatchStageSchema>;

export const TowSettlementStatusSchema = z.enum([
  "none",
  "pre_auth_holding",
  "final_charged",
  "refunded",
  "cancelled",
]);
export type TowSettlementStatus = z.infer<typeof TowSettlementStatusSchema>;

// ───────── Value objects ─────────

export const TowFareQuoteSchema = z.object({
  mode: TowServiceModeSchema,
  base_amount: z.number(),
  distance_km: z.number(),
  per_km_rate: z.number(),
  urgency_surcharge: z.number(),
  buffer_pct: z.number(),
  cap_amount: z.number(),
  locked_price: z.number().nullable().default(null),
  currency: z.string().default("TRY"),
});
export type TowFareQuote = z.infer<typeof TowFareQuoteSchema>;

export const TowLiveLocationSchema = z.object({
  case_id: z.string(),
  technician_id: z.string(),
  lat: z.number(),
  lng: z.number(),
  heading: z.number().nullable().default(null),
  speed_kmh: z.number().nullable().default(null),
  captured_at: z.string(),
});
export type TowLiveLocation = z.infer<typeof TowLiveLocationSchema>;

export const TowKaskoDeclarationSchema = z.object({
  has_kasko: z.boolean().default(false),
  insurer_name: z.string().optional(),
  policy_number: z.string().optional(),
  pre_auth_on_customer_card: z.boolean().default(true),
});
export type TowKaskoDeclaration = z.infer<typeof TowKaskoDeclarationSchema>;

export const TowDispatchAttemptSchema = z.object({
  id: z.string(),
  technician_id: z.string(),
  technician_name: z.string(),
  attempt_order: z.number().int(),
  sent_at: z.string(),
  response_at: z.string().nullable().default(null),
  response: z.enum(["pending", "accepted", "declined", "timeout"]).default("pending"),
  distance_km: z.number(),
  eta_minutes: z.number().int(),
});
export type TowDispatchAttempt = z.infer<typeof TowDispatchAttemptSchema>;

export const TowTechnicianProfileSchema = z.object({
  id: z.string(),
  name: z.string(),
  rating: z.number(),
  completed_jobs: z.number().int(),
  plate: z.string(),
  truck_model: z.string(),
  equipment: TowVehicleEquipmentSchema,
  phone: z.string(),
  photo_url: z.string().nullable().default(null),
});
export type TowTechnicianProfile = z.infer<typeof TowTechnicianProfileSchema>;

export const TowBidSchema = z.object({
  id: z.string(),
  technician: TowTechnicianProfileSchema,
  price_amount: z.number(),
  price_label: z.string(),
  eta_window_label: z.string(),
  equipment: TowVehicleEquipmentSchema,
  guarantee_label: z.string().nullable().default(null),
  submitted_at: z.string(),
});
export type TowBid = z.infer<typeof TowBidSchema>;

export const TowEvidenceKindSchema = z.enum([
  "customer_pre_state",
  "tech_arrival",
  "tech_loading",
  "tech_delivery",
]);
export type TowEvidenceKind = z.infer<typeof TowEvidenceKindSchema>;

export const TowEvidenceSchema = z.object({
  id: z.string(),
  case_id: z.string(),
  kind: TowEvidenceKindSchema,
  uploader: z.enum(["customer", "technician", "system"]),
  photo_url: z.string(),
  caption: z.string().nullable().default(null),
  created_at: z.string(),
});
export type TowEvidence = z.infer<typeof TowEvidenceSchema>;

export const TowOtpChallengeSchema = z.object({
  code: z.string(),
  purpose: z.enum(["arrival", "delivery"]),
  recipient: z.enum(["customer", "delivery_recipient"]),
  issued_at: z.string(),
  expires_at: z.string(),
  verified_at: z.string().nullable().default(null),
});
export type TowOtpChallenge = z.infer<typeof TowOtpChallengeSchema>;

// ───────── Request draft ─────────

export const TowRequestSchema = z.object({
  mode: TowServiceModeSchema,
  pickup_lat_lng: LatLngSchema.nullable().default(null),
  pickup_label: z.string(),
  dropoff_lat_lng: LatLngSchema.nullable().default(null),
  dropoff_label: z.string().nullable().default(null),
  vehicle_id: z.string(),
  incident_reason: TowIncidentReasonSchema,
  required_equipment: TowVehicleEquipmentSchema,
  scheduled_at: z.string().nullable().default(null),
  fare_quote: TowFareQuoteSchema,
  kasko: TowKaskoDeclarationSchema.default({
    has_kasko: false,
    pre_auth_on_customer_card: true,
  }),
  attachments: z.array(z.string()).default([]),
});
export type TowRequest = z.infer<typeof TowRequestSchema>;

// ───────── Case snapshot (mock/runtime) ─────────

export const TowCaseSnapshotSchema = z.object({
  id: z.string(),
  created_at: z.string(),
  updated_at: z.string(),
  request: TowRequestSchema,
  stage: TowDispatchStageSchema,
  assigned_technician: TowTechnicianProfileSchema.nullable().default(null),
  current_location: LatLngSchema.nullable().default(null),
  route_points: z.array(LatLngSchema).default([]),
  eta_minutes: z.number().int().nullable().default(null),
  dispatch_attempts: z.array(TowDispatchAttemptSchema).default([]),
  bids: z.array(TowBidSchema).default([]),
  accepted_bid_id: z.string().nullable().default(null),
  evidence: z.array(TowEvidenceSchema).default([]),
  otp_challenges: z.array(TowOtpChallengeSchema).default([]),
  settlement_status: TowSettlementStatusSchema.default("none"),
  final_amount: z.number().nullable().default(null),
  cancellation_reason: z.string().nullable().default(null),
  cancellation_fee: z.number().nullable().default(null),
  rating: z.number().nullable().default(null),
  review_note: z.string().nullable().default(null),
});
export type TowCaseSnapshot = z.infer<typeof TowCaseSnapshotSchema>;

// ───────── Helpers ─────────

export function computeTowCap(input: {
  base: number;
  distance_km: number;
  per_km: number;
  urgency_surcharge: number;
  buffer_pct: number;
}): number {
  const raw =
    input.base + input.distance_km * input.per_km + input.urgency_surcharge;
  return Math.round(raw * (1 + input.buffer_pct));
}

export function computeTowCancellationFee(
  mode: TowServiceMode,
  stage: TowDispatchStage,
): number {
  if (mode === "immediate") {
    if (stage === "searching") return 0;
    if (stage === "accepted" || stage === "en_route" || stage === "nearby")
      return 75;
    if (stage === "arrived") return 300;
    if (stage === "loading" || stage === "in_transit") return -1; // full fare
    return 0;
  }
  // scheduled — gerçek hesap için schedule_at - now farkı gerekir,
  // mock için bucket bazlı varsayılan:
  if (stage === "bidding_open" || stage === "scheduled_waiting") return 0;
  if (stage === "accepted" || stage === "en_route") return 150;
  if (stage === "arrived") return -1;
  return 0;
}

export const TOW_DEFAULT_QUOTE_PARAMS = {
  base: 950,
  per_km: 70,
  urgency_surcharge: 80,
  buffer_pct: 0.1,
} as const;
